ndc-lambda-sdk/src/inference.ts

```diff
@@ -105,27 +125,42 @@ export function printFunctionIssues(functionIssues: FunctionIssues) {
 type TypeDerivationContext = {
   objectTypeDefinitions: schema.ObjectTypeDefinitions
   scalarTypeDefinitions: schema.ScalarTypeDefinitions
+  customTypeNameRegistry: CustomTypeNameRegistry
   typeChecker: ts.TypeChecker
-  functionsFilePath: string,
   ndcLambdaSdkModule: ts.ResolvedModuleFull,
 }
 
+/** Provides a source for where the type was discovered. */
+type TypeSource = FunctionReturnTypeSource | SymbolTypeSource
+
+/** The type came from the return type of a function signature */
+type FunctionReturnTypeSource = {
+  type: "FunctionReturn",
+  callSignature: ts.Signature
+}
+
+/** The type came from something with a symbol (eg. function parameter, object property) */
+type SymbolTypeSource = {
+  type: "Symbol"
+  symbol: ts.Symbol
+}
+
 function cloneTypeDerivationContext(context: TypeDerivationContext): TypeDerivationContext {
   return {
     objectTypeDefinitions: structuredClone(context.objectTypeDefinitions),
     scalarTypeDefinitions: structuredClone(context.scalarTypeDefinitions),
+    customTypeNameRegistry: context.customTypeNameRegistry.clone(),
```

Comment:
Why is this cloned?

Reply:
Because this cloning function is used in places (relaxed types) where we run the type derivation algorithm but throw away the results, so we clone the state and discard it after.